packages/devtools_app/lib/src/flutter/table.dart

```diff
@@ -0,0 +1,395 @@
+import 'dart:math';
+
+import 'package:flutter/material.dart';
+
+import '../table_data.dart';
+import '../trees.dart';
+
+/// A table that shows [TreeNode]s.
+///
+/// This takes in a collection of [data], and a collection of [ColumnData].
+/// It takes at most one [TreeColumnData].
+///
+/// The [ColumnData] gives this table information about how to size its
+/// columns, how to present each row of `data`, and which rows to show.
+///
+/// To lay the contents out, this table's [_TreeTableState] creates a flattened
+/// list of the tree hierarchy. It then adds the nesting level of the
+/// deepest row in [data] that is currently visible with the prescribed
+/// width of each row in the table.
+class TreeTable<T extends TreeNode<T>> extends StatefulWidget {
+  const TreeTable({
+    Key key,
+    @required this.columns,
+    this.treeColumn,
+    @required this.data,
+    @required this.keyFactory,
+  }) : super(key: key);
+
+  /// The columns to show in this table.
+  final List<ColumnData<T>> columns;
+
+  /// The column of the table to treat as expandable.
+  final TreeColumnData<T> treeColumn;
+
+  /// The tree structure of rows to show in this table.
+  final T data;
+
+  /// Factory that creates keys for each row in this table.
+  ///
+  /// It uses [PageStorageKey] to restore the table's state when re-opening
+  /// its page.
+  final String Function(T data) keyFactory;
+
+  @override
+  _TreeTableState<T> createState() => _TreeTableState<T>();
+}
+
+class _TreeTableState<T extends TreeNode<T>> extends State<TreeTable<T>>
+    with TickerProviderStateMixin {
+  List<T> _flattenedList = [];
+
+  /// The width of each column in this table, including the
+  /// appropriate indentation for nested rows.
+  ///
+  /// This is computed in [_computeTableWidth].
+  List<double> columnWidths = [];
+  double get tableWidth => columnWidths.reduce((x, y) => x + y);
+
+  AnimationController _resizeController;
+  Animation<double> _resizeAnimation;
+  Tween<double> _widthTween;
+
+  /// The width to assume for columns that don't specify a width.
+  static const defaultColumnWidth = 500.0;
+
+  /// The maximum height to allow for rows in the table.
+  static const defaultRowHeight = 42.0;
+
+  @override
+  void initState() {
+    super.initState();
+    widget.data.expandCascading();
+    _flattenedList = _flattenExpandedTree();
```

Comment:
Why not just set `_flattenedList = _flattenExpandedTree(_flattenExpandedTree(filter: (n) => n.isExpanded);` from _computeTableWidth instead of calling it here as well?

Reply:
Lets add a TODO to revisit the logic for whether the _flattenedList should contain all nodes our just nodes that are currently visible.

Reply:
>  Why not just set `_flattenedList` from _computeTableWidth instead of calling it here as well?

To animate elements of the list in and out, we need to know what data to populate them with.  When we collapse a row, we update the data model immediately, but we don't complete animating the corresponding widget away for a few hundred milliseconds.

Keeping this whole list in memory is no more expensive than keeping the tree it came from, and it saves a lot of processing time and state management.

Reply:
yes this is good enough. it may become a little bit of a bottlekneck once we are talking 20000+ widgets for a large app.
The difference is the whole node tree doesn't have to be traversed on every set state but this will.

Reply:
This section here is only called on initState or on a widget rebuild.  The setState logic is doing computation on the column widths.

On widget rebuilds, we can check if the data lists have changed.

Now that we're calculating the column widths for the entire list instead of only the non-expanded items, we can actually do this computation only on initState and widget updates.